src/v3/src/transformer/field/transform.ts

```diff
@@ -41,28 +49,72 @@ const mapUiElement = (input: Input): FieldElement => {
   };
 };
 
-const getValidationMessages = (
-  fieldName: string,
+const getValidationMessages: ValidationErrorTransformer = (
+  input: Input,
+  data: Record<string, unknown>,
   widgetProps: WidgetProps,
   step?: NextStep,
-): IdxMessage[] => {
+): WidgetMessage[] | undefined => {
+  const { name: fieldName, type, maxLength } = input;
   const { name } = step || {};
-  const errorMessage: IdxMessage = {
-    class: 'ERROR',
-    message: loc('model.validation.field.blank', 'login'),
-    i18n: { key: 'model.validation.field.blank' },
-  };
+  const fieldValue = data[fieldName];
+
+  // Customized i18n key support for fields
   const customizedErrorConfig = [
     { field: 'identifier', key: 'error.username.required' },
     { field: 'credentials.passcode', key: 'error.password.required' },
   ].find((obj) => obj.field === fieldName);
-  if (customizedErrorConfig
-    && name === IDX_STEP.IDENTIFY
-    && isCustomizedI18nKey(customizedErrorConfig.key, widgetProps)) {
-    errorMessage.message = loc(customizedErrorConfig.key, 'login');
-    errorMessage.i18n.key = customizedErrorConfig.key;
+  const useCustomizedBlankErrorMessage = !!customizedErrorConfig && name === IDX_STEP.IDENTIFY
+                                    && isCustomizedI18nKey(customizedErrorConfig.key, widgetProps);
+
+  const populatedFieldErrorChecks: ValidationErrorTester[] = [
+    {
+      // Validate maxLength for string inputs that have the property set
+      tester: (value?: unknown) => !!maxLength && type === 'string' && !!value && (value as string).length > maxLength,
+      message: () => (
+        {
+          class: 'ERROR',
+          message: loc('model.validation.field.string.maxLength', 'login'),
+          i18n: { key: 'model.validation.field.string.maxLength' },
+        }
+      ),
+    },
+  ];
+  const blankFieldErrorChecks: ValidationErrorTester[] = [
+    {
+      tester: () => useCustomizedBlankErrorMessage,
+      message: () => (
+        {
+          class: 'ERROR',
+          // As long as message() is only ever called after tester(), we can be sure that customizedErrorConfig is defined
+          message: loc(customizedErrorConfig!.key, 'login'),
+          i18n: { key: customizedErrorConfig!.key },
+        }
+      ),
+    },
+    {
+      // Default blank field error message should be rendered as long as no custom i18n keys are set in widget config
+      tester: () => !useCustomizedBlankErrorMessage,
+      message: () => (
+        {
+          class: 'ERROR',
+          message: loc('model.validation.field.blank', 'login'),
+          i18n: { key: 'model.validation.field.blank' },
+        }
+      ),
+    },
+  ];
+
+  let messages: WidgetMessage[] | undefined;
+  // Validation errors can be divided into populated and blank field checks
+  if (fieldValue) {
```

Comment:
Would like some thoughts on whether populated error checks vs. blank error checks is a good (and exhaustive) split for this error testing pattern? Per description, see `transformUnhandledErrors.ts` for the pattern that this logic was modeled on